### ISSUE TITLE
Node locations are not saved

### ISSUE THREAD

**Issue by steindani** [shapesecurity/shift-java#104]
#102 introduced a bug: the locations were calculated, but not stored.

Since `com.shapesecurity.shift.parser.ParserWithLocation#locations` was not replaced with the result of `locations.put` and also it was marked as `final`, the locations collection was not growing.

**Comment by steindani** [comment]
Please also consider making `locations` public.

In _my usecase_ I don't want to use the parser more than once. I had problems with separate source codes with matching `BindingIdentifier`s resulting in the same Java `Object` and this was not an expected result. So I wouldn't mind if I did not have to pass the whole parser object forward, only the location information.

**Comment by michaelficarra** [comment]
Haha yeah, I expected this to be horribly broken right now. I will be adding location tests soon. Go ahead and send a PR to make the locations table public, but I'm not sure yet how I feel about that.

**Comment by bakkot** [comment]
@steindani, if we don't end up making it public (or in the meantime), you can always just subclass `ParserWithLocation`; something like:

``` java
public class ParserExposingLocationTable extends ParserWithLocation {
    public HashTable<Node, SourceSpan> getLocationTable() {
         return locations;
    }
}
```

**Comment by steindani** [comment]
Thanks!

– Right now I'm fine with passing the parser, but this might need a better solution. It's not a burning problem. (Maybe a getter, since this way the variable couldn't be set from the outside, but the current "structure" could be acquired?)

– Is it the expected behaviour from the parser to result in the same Java Object for the same `BindingIdentifier` after using the same parser on two different sources? Is there some kind of caching?

– I did not have time yet to test it out, but the locations might have a one-off error. I've noticed it with the `FunctionDeclaration`s, the start location is always marking the previous (empty) line. The end location seems to be fine.

**Comment by steindani** [comment]
Oh, right. Thanks, @bakkot!

**Comment by bakkot** [comment]
They shouldn't be giving you the same object, but we've overridden equality on nodes. So if you're comparing using `.equals`, two nodes with identical properties will compare appear as equal.

The locations table is actually using equality instead of identity, which is a bug; I've fixed it in #106.

**Comment by steindani** [comment]
Yeah, I've made the same mistake, instead of using `IdentityHashMap` I've
used `HashMap` in my code while converting the AST. Thanks for pointing
out!
